sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +501,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
         // not(not(e))  =>  e
         case Not(e) => e
+        // De Morgan's law: !(a || b) => !a && !b
```

Comment:
I mean we couldn't tell `And(Not(lhs), Not(rhs))` is more optimal than `Not(Or(lhs, rhs))`, do we? Sorry if I missed something.